xla/backends/profiler/gpu/cupti_tracer.h

```diff
@@ -48,6 +72,207 @@ struct CuptiTracerOptions {
+};
+
+// Container for PM sampling decode thread
+// Responsible for fetching PM sampling data from device and providing it to
+// handler or other external container
+class PmSamplingDecodeThread {
+  enum thd_state {
+    // Thread is starting, not yet ready to be enabled
+    state_uninitialized_,
+    // Thread is ready for enablement but decoding has not yet been triggered
+    state_initialized_,
+    // Thread is enabled, polling for metrics from all devices
+    state_enabled_,
+    // Thread is disabled, not polling for metrics, but could be re-enabled
+    state_disabled_,
+    // Thread is finishing and guaranteed to return, allowing join
+    state_exiting_
+  };
+
+  // Current state of the thread (only set by worker thread)
+  volatile thd_state current_thd_state_ = state_uninitialized_;
+    
+  // State thread should transition to (only set by main thread)
+  volatile thd_state nextThdState_ = state_initialized_;
+
+  // Thread handle
+  std::thread* thd_;
+
+  // Function run by thd_
+  static void ThdFunc(PmSamplingDecodeThread* control);
+
+  // Isolate the main decode loop
+  static void ThdFuncDecodeUntilDisabled(PmSamplingDecodeThread* control);
+
+  // Devices to decode by this thread
+  std::vector< std::shared_ptr<PmSamplingDevice> > devs_;
+
+  public:
+  // Spin wait sleep period, set to the min of this and all device periods
+  // Space to asynchronously initialize this class and the thread it spawns
+  absl::Duration decode_period_ = absl::Seconds(1);
+  PmSamplingDecodeThread(std::vector< std::shared_ptr<PmSamplingDevice> >
+      devs);
+
+  // Signal thread to exit; join thread
+  ~PmSamplingDecodeThread() {
+    nextThdState_ = state_exiting_;
+    thd_->join();
+  }
+
+  // Signal and test for state transitions
+  bool IsThdInitialized() { return current_thd_state_ == state_initialized_; }
+  void ThdIsInitialized() { current_thd_state_ = state_initialized_; }
+
+  void EnableThd() { nextThdState_ = state_enabled_; }
+  bool ShouldThdEnable() { return nextThdState_ == state_enabled_; }
+  void ThdIsEnabled() { current_thd_state_ = state_enabled_; }
+  bool IsThdEnabled() { return current_thd_state_ == state_enabled_; }
+
+  void DisableThd() { nextThdState_ = state_disabled_; };
+  bool ShouldThdDisable() { return nextThdState_ == state_disabled_; }
+  void ThdIsDisabled() { current_thd_state_ = state_disabled_; }
+  bool IsThdDisabled() { return current_thd_state_ == state_disabled_; }
+
+  void ExitThd() { nextThdState_ = state_exiting_; }
+  bool ShouldThdExit() { return nextThdState_ == state_exiting_; }
+  void ThdIsExiting() { current_thd_state_ = state_exiting_; }
+  bool IsThdExiting() { return current_thd_state_ == state_exiting_; }
+};
+
+// Should be safe on all hardware
+struct PmSamplingConfig {
+  // Whether to enable PM sampling
+  bool enable_pm_sampling = false;
+  // List of metrics to enable
+  std::vector<const char*> metrics{};
```

Comment:
```suggestion
  std::vector<std::string> metrics;
```
Let's use std::string here it makes it easier to reason about ownership.